examples/lstm.py

```diff
@@ -0,0 +1,255 @@
+        f_t = sigmoid((x_t @ self.W_f) + (h_tm1 @ self.R_f) + self.b_f)
+        o_t = sigmoid((x_t @ self.W_o) + (h_tm1 @ self.R_o) + self.b_o)
+
+        c_hat_t = numpy.tanh(
+            (x_t @ self.W_c) + (h_tm1 @ self.R_c) + self.b_c
+        )
+
+        c_t = f_t * c_tm1 + i_t * c_hat_t
+        h_t = o_t * numpy.tanh(c_t)
+
+        return h_t, c_t
+
+    def apply(self, x):
+        """Apply the layer."""
+        s = self.s0
+        c = self.c0
+        for e in x:
+            s, c = self.step(e, s, c)
+        # Maybe collect and return the full list of s outputs?
+        return s
+
+
+@dataclass(frozen=True)
+class Sequential(ArithmeticData):
+    """Sequential layer, applies all sub-layers in order."""
+
+    layers: 'Tuple of layers'
+
+    def apply(self, x):
+        """Apply the layer."""
+        for layer in self.layers:
+            x = layer.apply(x)
+        return x
+
+
+def cost(model, x, target):
+    """Square difference loss."""
+    y = model.apply(x)
+    diff = target - y
+    return sum(diff * diff)
+
+
+# @myia(backend_options={'target': device_type})
+@myia(backend='pytorch', backend_options={'device': device_type})
+def step(model, x, y):
+    """Returns the loss and parameter gradients."""
+    # value_and_grad will return cost(model, x, y) and dcost(...)/dmodel.
+    # The 'model' argument can be omitted: by default the derivative wrt
+    # the first argument is returned.
+    return value_and_grad(cost, 'model')(model, x, y)
+
+
+def run_helper(epochs, n, batch_size, layer_sizes):
+    """Run a model with the specified layer sizes on n random batches.
+
+    The first layer is an LSTM layer, the rest are linear+tanh.
+
+    Arguments:
+        epochs: How many epochs to run.
+        n: Number of training batches to generate.
+        batch_size: Number of samples per batch.
+        layer_sizes: Sizes of the model's layers.
+    """
+    layers = []
+    lstmp, *linp = lstm_parameters(*layer_sizes, batch_size=batch_size)
+    layers.append(LSTMLayer(*lstmp))
+    for W, b in linp:
+        layers.append(Linear(W, b))
+        layers.append(Tanh())
+    model = Sequential(tuple(layers))
+    data = generate_data(n, batch_size, layer_sizes[0], layer_sizes[-1], 10)
+    lr = getattr(numpy, dtype)(0.01)
+
+    for _ in range(epochs):
+        costs = []
+        t0 = time.time()
+        for inp, target in data:
+            cost, dmodel = step(model, inp, target)
+            if isinstance(cost, numpy.ndarray):
+                cost = float(cost)
+            costs.append(cost)
+            model = model - (lr * dmodel)
+        c = sum(costs) / n
+        t = time.time() - t0
+        print(f'Cost: {c:15.10f}\tTime: {t:15.10f}')
+
+
+# def test_run():
+#     """Run the model.
+
+#     This function is run automatically in the test suite to check against
+#     regressions, so keep a low number of narrow layers to make sure it runs
+#     quickly.
+#     """
+#     run_helper(1, 1, 5, (10, 3))
```

Comment:
Maybe a comment to explain that we don't test LSTM yet.  Or better, make it fail with xfail.

Reply:
OK